Updater component

rmkraus opened the pull request:
1) Added updater component
2) Updated the update script to better work with the updater component
3) Added card and more-info for updater to the frontend in [home-assistant-polymer PR #1](https://github.com/balloob/home-assistant-polymer/pull/1)
4) Rebuilt frontend using new card and more-info

rmkraus committed:
Merged in Updater component.

rmkraus committed:
Frontend Rebuild

Rebuilt fronted using version from [home-assistant-polymer PR
#1](https://github.com/balloob/home-assistant-polymer/pull/1)

rmkraus committed:
Update script updates

rmkraus committed:
Quit signal handling

Added handler for SIGQUIT to safely exit HA

rmkraus committed:
Fixed typo when binding to SIGQUIT

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+# homeassistant constants
+DOMAIN = "updater"
+DEPENDENCIES = []
+REQUIREMENTS = ['requests>=2.0', 'gitpython>=1.0.1']
```
No need to depend on `requests`, that is a core dependency.

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+CONF_LOG_FILE = "log_file"
+CONF_REPO_NAME = "repo_name"
+CONF_REPO_BRANCH = "repo_branch"
+HA_SOURCE_DIR = os.path.abspath(os.path.join(homeassistant.__file__,
```
The correct way to get the directory of a file is using `os.path.dirname`:
```python
HA_SOURCE_DIR = os.path.abspath(os.path.join(os.path.dirname(homeassistant.__file__), '..'))
```

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+    @property
+    def state_attributes(self):
+        ''' current entity state attributes '''
+        return {'hidden': self.hidden,
```
No need to set `hidden`. Overwriting the property ensures it is set.

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+    Setup updater component.
+    """
+    # get optional configuration
+    pid_file = config['updater'].get(CONF_PID_FILE, None)
```
Could you use `DOMAIN` for the key?

```python
pid_file = config[DOMAIN].get(CONF_PID_FILE, None)
```

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+HA_SOURCE_DIR = os.path.abspath(os.path.join(homeassistant.__file__,
+                                             '..', '..'))
+GH_API_CALL = 'https://api.github.com/repos/{repo}/commits?sha={branch}'
+SCAN_INTERVAL = 60 * 60 * 12
```
Looks like this constant is never used.

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+            self.versions['newest_url'] = github_data[0]['html_url']
+
+        # find local copy info
+        if self._repo_class is not None:
```
The local info will never change. We should be able to move this logic to the `setup` method and pass `current_sha` as a parameter to the Updater class.

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+            self._repo_class = Repo
+
+        self.hass = hass
+        self._logger = logger
```
You don't seem to be using this variable

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+
+        self.config = kwargs
+
+        hass.track_time_change(self.update, hour=0, minute=0, second=0)
```
Please wrap this method in a function defined in the line above. The advantage of this is that the function will be named which can help us a lot if we ever have to debug it. This has added benefit that the method signature of `update` can be reduced to `def update(self)`.

```python
def check_ha_update(event):
    self.update()

hass.track_time_change(check_ha_update, hour=0, minute=0, second=0)
```

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+    entity_id = 'updater.updater'
+    name = 'Updater'
+
+    def __init__(self, hass, logger, **kwargs):
```
We always know how this class is being called, no need to add `**kwargs``. Only use the catch-alls when you need to

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+    repo_branch = config['updater'].get(CONF_REPO_BRANCH, 'master')
+
+    # create component entity
+    Updater(hass, _LOGGER, repo_name=repo_name, branch=repo_branch,
```
No need to pass _LOGGER in. It is a constant in the global scope and does not need to be passed around.

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+                'date': self.versions['newest_date'],
+                'link': self.versions['newest_url']}
+
+    def run_update(self, *args, **kwargs):
```
You can rewrite this to `def run_update(self, service_call)`

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+
+    def run_update(self, *args, **kwargs):
+        ''' run the updater '''
+        upscript = os.path.join(HA_SOURCE_DIR, 'scripts', 'update')
```
This method does not depend on anything specific to the updater entity. It would be cleaner if you would extract the registering of the service and this service method to be done inside `setup`.

balloob commented on scripts/update:
```diff
@@ -1,8 +1,63 @@
+# PID - The Process ID of Home Assistant if it is currently running. If this
+#       is supplied, Home Assistant will be shutdown before updating and then
+#       restarted after the update has been applied.
+# LOG_FILE - A file where the information usually printed to the screen should
```
The LOG_FILE and PID_FILE logic doesn't really belong in an update script. I think this should be moved to a separate LAUNCH_DAEMON script.

We could have the `run_update` service callback decide if it should launch the regular update script or run LAUNCH_DAEMON with an update flag set.

rmkraus replied:
I think there are two ways to accomplish this.

1) Is to create a Daemon Helper component that will manage PID and Log files. This seems a bit heavy for such a simple feature.
2) Create PID and LOG flags that HA will interpret in __main__ when launching.

I can't think of any other way we could have the locations of the PID and Log files accessible from inside HA, but there may very well be something I'm not considering. I personally would rather the second option.

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+        self.hass = hass
+        self._logger = logger
+
+        self.config = kwargs
```
I would prefer if you write it out:
```python
self.repository = repository
self.branch = branch
```

balloob commented:
We should make sure that balloob/home-assistant-polymer#1 is merged before we merge this PR so the `home-assistant-polymer` submodule is pointing at the correct sha.

rmkraus committed:
Applied suggested fixes to Updater component

Applied the suggested fixes to the Updater component, recompiled the
frontend, updated subproject home-assistant-polymer to latest master
version.

rmkraus committed:
pylint fixes to Updater

rmkraus committed:
Built in daemon functionality

Daemon functionality is built in. —verbose to enable verbose logging.
—pid-file to create and check pid file. —daemon to run HA as daemon.

rmkraus commented:
I made the daemon functionality built into HA with command line flags. Checkout __init__, __main__, and bootstrap to let me know what you think. This isn't ready for merging yet. If we like these changes, then I'll have to apply the changes to the update module and script to use these flags.

rmkraus committed:
Removed unused import

balloob commented on homeassistant/__init__.py:
```diff
@@ -45,6 +46,8 @@
 class HomeAssistant(object):
     """ Core class to route all communication to right components. """
 
+    arguments = None
```
I don't see this being used.

rmkraus replied:
It is set on line 212 of __main__. It is meant to propagate cmd line arguments into HA so they can be recalled later. Like in the updater component.

balloob replied:
As we discussed on the chat, we need a way to get those arguments to the updater so it can call the update script with the right parameters.

I came up with two more solutions:
 1. Call ArgumentParser from updater component. Yes, will not work with other components, if the moment arrives that we ever need it, we can switch to solution 2.
 2. Have \_\_main\_\_ setup a restart method in a util package that has all parameters available

Example of 2:
```python
def restart_ha():
    […]

import homeassistant.util.restart as restart

restart.restart_ha = restart_ha
```

balloob commented on homeassistant/__main__.py:
```diff
@@ -109,10 +109,71 @@ def get_arguments():
+        sys.exit(0)
+
+
+def setup_daemon(pid_file, run_as_daemon):
```
I think that this function can be broken into three parts:

 1. kill existing HA if pid given
 2. Daemonize if requested
 3. Save pid if pid is given

rmkraus committed:
Fixed problem with verbose logging. Cleaned up STDOUT.

Cleaned a few caught instances where data was being “logged” without
being sent through a logger instance. Fixed bug with verbose logging
still not being verbose.

balloob commented on homeassistant/__main__.py:
```diff
@@ -127,16 +188,19 @@ def main():
-        })
+        hass = bootstrap.from_config_dict(
+            {http.DOMAIN: {}, demo.DOMAIN: {}},
+            args=args)
```
Please only sent minimum required information through. In this case `daemon=True/False`, `verbose=True/False`. That the data came from parsing args is an implementation detail of `__main__.py`.

balloob commented on homeassistant/bootstrap.py:
```diff
@@ -199,30 +200,32 @@ def from_config_file(config_path, hass=None):
-        ))
-    except ImportError:
-        _LOGGER.warn("Colorlog package not found, console coloring disabled")
+    if not daemon:
```
We should move this check to the function calling `enable_logging` and have it not call `enable_logging` instead.

rmkraus replied:
It still needs to call enable_logging because we still want the log file.

balloob replied:
Oh missed the second part. Guess we can split up the method in two parts to keep each method do one thing*

*: I've been doing a lot of functional programming lately, tell me if I go a bit overboard…

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+    try:
+        from git import Repo
+    except ImportError:
+        _LOGGER.error('Missing package gitpython')
```
We should return `False` to mark loading this component as failed.

rmkraus replied:
Fixed.

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+        self.update_ha_state()
+
+    @property
+    def hidden(self):
```
It makes more sense if the `state` property depends on the `hidden` property.

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+        hass.track_time_change(check_ha_update, hour=0, minute=0, second=0)
+
+        # initialize data
+        self.versions = {'newest_sha': None, 'newest_msg': None,
```
Is there a reason why you're using a hash instead of `self.newest_sha` etc?

rmkraus replied:
Specifically because pylint was yelling at me for having too many properties. It is kind of 6 one way, half dozen the other. I figured I would err on the side of not using another pylint exception.

balloob replied:
I prefer to disable the pylint exception over using hashes to work around it. It is a bit confusing now

rmkraus committed:
pylint fixes, windows check for daemons

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+            _LOGGER.warning('Error communicating with GitHub')
+
+        # update with HA
+        self.update_ha_state()
```
I don't think we should update the state of the entity if we had an error communicating with GitHub.

balloob commented on homeassistant/components/updater.py:
```diff
@@ -0,0 +1,143 @@
+    @property
+    def hidden(self):
+        ''' only show the entity when an update is available '''
+        return not self.versions['newest_sha'] and \
```
If you fail the setup of the component if `current_sha` cannot be detected, you can simplify this check with `return self.newest_sha != self.current_sha`

balloob commented on homeassistant/components/history.py:
```diff
@@ -147,7 +147,8 @@ def _api_history_period(handler, path_match, data):
 
-    print("Fetchign", start_time, end_time)
+    # print("Fetchign", start_time, end_time)
+    # Should this be a call to a logger?
```
Debug comments

balloob commented on homeassistant/components/http.py:
```diff
@@ -202,6 +202,11 @@ def register_path(self, method, url, callback, require_auth=True):
         """ Regitsters a path wit the server. """
         self.paths.append((method, url, callback, require_auth))
 
+    def log_message(self, fmt, *args):
```
I don't get why these are here. It would be highly preferred if we don't have to spread the fact that we route our log to a file throughout the app.

rmkraus replied:
log_message is the function that the BaseHTTP objects have built in to write their output to the screen. This forces the HTTP Server to use the same logging as has been configured for the rest of the application. If we don't, then HTTP Server messages will always be written to the screen. Basically, this implements the functionality that the built-in HTTP Server objects should have built in, but don't.

As an aside, this is also the reason that the HTTP Server messages are always formatted differently form all the other messages.

balloob replied:
I get it now, they are overriding the log methods of the built-in classes. Could we add a comment to not confuse future readers of the code?

balloob commented on homeassistant/components/http.py:
```diff
@@ -219,6 +224,10 @@ def __init__(self, req, client_addr, server):
         self._session = None
         SimpleHTTPRequestHandler.__init__(self, req, client_addr, server)
 
+    def log_message(self, fmt, *arguments):
```
Same as above.

rmkraus committed:
Applied suggestions to daemon functions and updater

rmkraus committed:
Removed debugging lines from history

rmkraus commented:
I'm closing this as it is really far behind.
